content/posts/2021-11-13---Typy.md

```diff
@@ -0,0 +1,286 @@
+class Figure {
+    int side;
+    int height;
+}
+class Triangle extends Figure {
+    int area = (side*height)/2;
+}
+```
+Mamy na początku daną klasę `Figure`, zawierającą 2 pola: bok i wysokość.
+Następnie tworzymy klasę reprezentującą trójkąt, która ma zmienną o wartości pola danego trójkąta.
+Widzimy, że klasa `Triangle` nie ma własnych pól, tylko korzysta z pól klasy `Figure`. Wyboraźmy sobie sytuację, kiedy mamy klasy wszystkich figur geometrycznych.
+Wiele z nich zawiera te same pola, więc nie ma sensu tworzenia ich dla każdej klasy osobno. Dzięki dziedziczeniu klas mamy możliwość skorzystania z pól klasy nadrzędnej(poza prywatnymi),
+dla każdej z klas podrzędnych. I to jest właśnie cała magia dziedziczenia.
+Dziedziczenie z konkretnej klasy sygnalizujemy słowem "extends" - czyli rozszerza.
+W wolnym tłumaczeniu - "klasa trójkąt rozszerza klasę figury", w związku z tym ma dostęp do jej zawartości.
+Tworzenie konstruktora w klasie podrzędnej, wygląda tak jak zawsze, jednak gdy korzystamy z pól klasy nadrzędnej korzystamy w konstruktorze z metody `super()`, aby wywołać konstruktor klasy wyżej.
+
+### Interfejsy i ich implementacja
+O interfejsach wspominałem już wyżej, gdy mowa była o klasach abstrakcyjnych. W gruncie rzeczy są do nich dosyć podobne, jest jednak kilka różnic o których warto wspomnieć. 
+Interfejs to taki typ, który jedynie mówi nam, jakie metody należy zaimplementować w ramach danej klasy. Te metody nie posiadają ani ciała, ani nawiasów klamrowych.
+Tworzymy go podobnie jak zwykłą klasę jednak zamiast słowa "class", korzystamy ze słowa "interface", jak w poniższym przykładzie:
+```java
+public interface Code {
+    public void learnFromYoutube();
+    public int hoursToLearn();
+    public boolean isMyProjectFinnished();
+}
+```
+Jak widzimy, interfejs zawiera jedynie metody, które programista korzystający z danego interfejsu MUSI zaimplementować. Jest to w dużym skrócie po prostu gotowy szablon klasy, z której zamierzamy korzystać.
+Jest to niezwykle przydatne w praktyce, podczas planowania schematu naszego programu. Istotne jest to, że interfejsu mogą rozszerzać jedynie inne interfejsy. 
+Poza zwykłymi metodami, interfejsy mogą zawierać dodatkowo metody domyślne, prywatne, statyczne oraz stałe.
+Implementacja danego interfejsu jest również niezwykle prosta. Jedyne co należy zrobić, to zaimplementować metody tego interfejsu jak na poniższym przykładzie:
+```java
+    Code code = new Code() {
+        @Override
+        public void learnFromYoutube() {
+
+        }
+
+        @Override
+        public int hoursToLearn() {
+            return 0;
+        }
+
+        @Override
+        public boolean isMyProjectFinnished() {
+            return false;
+        }
+    };
+```
+Tutaj widzimy domyślnie wygenerowane przez interfejs metody do implementacji. Adnotacja override sygnalizuje nam, że musimy te metody nadpisać.
+Poniżej można zapoznać się z przykładową implementacją tych metod:
+```java
+    Code code = new Code() {
```

Comment:
Stwórz klase, która implementuje ten interfejs, bo to bardziej standardowe rozwiązanie niż przedstawiona przez ciebie anonimowa implementacja